processor/src/fast/mod.rs

```diff
@@ -375,18 +420,39 @@ impl FastProcessor {
                     }
                 },
                 Continuation::FinishJoin(node_id) => {
+                    self.check_extract_trace_state(
+                        NodeExecutionPhase::End(node_id),
+                        &mut continuation_stack,
+                        &current_forest,
+                    );
+
                     self.finish_join_node(node_id, &current_forest, host)?
                 },
                 Continuation::FinishSplit(node_id) => {
+                    self.check_extract_trace_state(
+                        NodeExecutionPhase::End(node_id),
+                        &mut continuation_stack,
+                        &current_forest,
+                    );
                     self.finish_split_node(node_id, &current_forest, host)?
                 },
                 Continuation::FinishLoop(node_id) => {
                     self.finish_loop_node(node_id, &current_forest, &mut continuation_stack, host)?
                 },
```

Comment:
Q: why don't we need to call `self.check_extract_trace_state(...)`?

Reply:
I thought I had fixed this - but apparently I had done it only for the `Start` ones. I moved all the `check_extract_trace_state()` calls inside the `finish()` methods to remove the confusion when looking at that part of the code.

But to answer your question directly - it's not that we don't call it, it's that we call it with a different `NodeExecutionPhase` depending on the `condition` at the top of the stack, so we're forced to call it inside `finish_loop_node()`.